### PR TITLE
Fix #1459: Added Post and Posts to NextJS

### LINKED ISSUE #1459: [next] Port Post and Posts to Next.js
**What would you like to be added**:
This issue shall address the porting of the mention items to Next.js
**Why would you like this to be added**:
Next.js migration

### PR DESCRIPTION
**Pull request by rogercyyu**
## Issue This PR Addresses

This issue address #1459.
 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [X] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This port `Post/` and `Posts/` to Next,js and converts it to TypeScript (most i think?).
* Tested it by commenting out things that aren't implemented yet. So works mostly.
* Requires a lot more components that haven't been implemented yet.
* Moved `telescope-post-content.css` to `\styles`.
* Probably some bugs yet to be discovered.

Thoughts? 

Thank you for your time and review.
## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/frontend/next/src/components/Post/Post.tsx
Comment (rogercyyu) [inline src/frontend/next/src/components/Post/Post.tsx:8]: Typo, needs to be placed imported in `index.tsx` as global to work.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (rogercyyu) [inline src/frontend/next/src/components/Post/Post.tsx:9]: Typo: This statement needs to be added to `index.tsx` to work properly.

**Comment by tonyvugithub** [comment]
@rogercyyu : Hi there, we are in the process of reviewing these migrated to Next PRs. Just wonder if you are still interested to work on this as we would like you to make some changes to your PR. Please let us know in case you decided to drop this then we can assign the issue to others. Thank you.

**Comment by rogercyyu** [comment]
@tonyvugithub I am happy to make some changes to this PR. Thanks

**Comment by tonyvugithub** [comment]
Got it. Thanks. We keep you posted about the progress.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:1]: You don't need to import `React` in next.js, but the other two are fine.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [reply src/frontend/next/src/components/Post/Post.tsx:9]: cc @tonyvugithub, can you take over the global styling we need?  We should get an issue filed to move the telescope-post-content.css file into our global css for next.js.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:11]: The next.js docs/examples on using TypeScript suggest using `type` vs. `interface` for Props.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:119]: We should file an issue to switch this to use `useSWR` instead.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:15]: This is really a `Post` vs. `DataItem`.  We probably should create something like this: https://github.com/vercel/next.js/tree/canary/examples/with-typescript/interfaces and move `interface Feed {...}` and `interface Post {...}` in there, then import them into this file so we can use them.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:25]: Typing needed on `theme`, see https://github.com/mui-org/material-ui/blob/master/examples/nextjs-with-typescript/src/ProTip.tsx#L15

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/components/Post/Post.tsx
Comment (tonyvugithub) [reply src/frontend/next/src/components/Post/Post.tsx:9]: @humphd : I believe @PedroFonsecaDEV is currently working on the global.css file in [#1548](https://github.com/Seneca-CDOT/telescope/issues/1548)

Comment (tonyvugithub) [reply src/frontend/next/src/components/Post/Post.tsx:9]: Also can we convert the css file of Post to JSS to keep it consistent with the rest of the components? And then we don't have to worry about moving the css to global.css

Comment (humphd) [reply src/frontend/next/src/components/Post/Post.tsx:9]: I don't think we can, or we can't do it easily.

The way we load and display posts now is to send raw HTML and inject it as `innerHTML`.  This means we can't rely on module scoped CSS for a component, since it isn't really a component.

It might be possible down the road to use something like [Server Components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html), but they are still very new.  https://addyosmani.com/blog/react-server-components/ has a good discussion.  This is a research problem we could tackle on the way to 2.0.

For now, we should probably port it as is.

Comment (humphd) [reply src/frontend/next/src/components/Post/Post.tsx:9]: OK, left a comment in #1548.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:15]: No.  The difference here between `Props`, which is done correctly, and this (which isn't) is that `Props` is internal to this code.  Whereas, we'll share the definition of `Feed` and `Post` throughout the frontend.

We need to create a folder at the same level as `components` and include an `index.ts` file, like this: https://github.com/vercel/next.js/tree/canary/examples/with-typescript/interfaces

Inside it, you can `export` two types:

```ts
export type Feed  = {
  link: string
  author: string
}

export type Post {
  feed: Feed
  id: string
  post: string
  title: string
  updated: string
  url: string
  html: string
}
```

Then you can import and use them here.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/components/Post/Post.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Post/Post.tsx:15]: Thank you for the detail feedback

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:4]: next.js won't let us do this, we'll have to load this globally, see https://nextjs.org/docs/basic-features/built-in-css-support.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:117]: It looks like you have both a fetch and a useSWR way in the same file.  Can we just do useSWR?

File: src/frontend/next/src/components/Posts/LoadAutoScroll.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/LoadAutoScroll.tsx:23]: See https://github.com/typescript-cheatsheets/react#forwardrefcreateref

File: src/frontend/next/src/components/Posts/Posts.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Posts.tsx:9]: I've got these finished and merged now, if you want to rebase.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Timeline.tsx:11]: This will be an Array of Post objects, so we should be able to type it with your interface.

File: src/frontend/next/src/interfaces/index.tsx
Comment (humphd) [inline src/frontend/next/src/interfaces/index.tsx:6]: Drop the `*Type` suffix.

File: src/frontend/next/src/pages/_app.tsx
Comment (humphd) [inline src/frontend/next/src/pages/_app.tsx:4]: These won't work here, needs to be globally imported.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/pages/_app.tsx
Comment (rogercyyu) [reply src/frontend/next/src/pages/_app.tsx:4]: Whoops, those are mistakes leftover from testing.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Post/Post.tsx:117]: Do you mean, change to: 
```typescript
  // Grab the post data from our backend so we can render it
  const fetcher = (url: string) => fetch(url).then((r) => r.json());

  const { data: post, error } = useSWR<Post>(postUrl, fetcher);
```
?

File: src/frontend/next/src/components/Posts/LoadAutoScroll.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Posts/LoadAutoScroll.tsx]: Not sure, if I fixed this correctly.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [reply src/frontend/next/src/components/Post/Post.tsx:10]: You can remove this comment now, @PedroFonsecaDEV is doing it in another PR in the global stylesheet.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:106]: We can't have 2 things named Post.  You are importing a `Post` type, and now defining another `Post` as the component.  You could call this `PostComponent` if you want.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:109]: We should type this: `useRef<HTMLSectionElement>();` I think.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:113]: Post vs. PostType

Comment (humphd) [reply src/frontend/next/src/components/Post/Post.tsx:113]: Also, I don't think we need this `fetcher` since it's the default.  You can drop the second arg and delete line 110, 111.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:121]: You need to import `AdminButtons` to use it here.  It's happening in #1515, which we need to land to get it here.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:125]: ErrorRoundedIcon is not imported.

File: src/frontend/next/src/components/Post/Post.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/Post.tsx:147]: You need Spinner, see #1479

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Timeline.tsx:1]: You can drop this.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Timeline.tsx:11]: ```Array<Post>```

File: src/frontend/next/src/pages/_app.tsx
Comment (humphd) [inline src/frontend/next/src/pages/_app.tsx:4]: cc @PedroFonsecaDEV, this is the file we need to have in our global CSS for syntax highlighting to work.

We can remove this from here.

File: src/frontend/next/src/styles/telescope-post-content.css
Comment (humphd) [inline src/frontend/next/src/styles/telescope-post-content.css:1]: This file can be removed, @PedroFonsecaDEV is doing it globally.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/pages/_app.tsx
Comment (PedroFonsecaDEV) [reply src/frontend/next/src/pages/_app.tsx]: Okay. Working on it.

Comment (PedroFonsecaDEV) [reply src/frontend/next/src/pages/_app.tsx]: @humphd done: 
Pull: https://github.com/Seneca-CDOT/telescope/pull/1579

**Comment by PedroFonsecaDEV** [comment]
This PR looks good. 
Still depending on #1515. 
Maybe a good option would be to create a simple `AdminButtons` on the right folder just to merge this one?
```
const AdminButtons = () => <h3> AdminButtons<h3>
```

File: src/frontend/next/src/components/Post/PostComponent.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/PostComponent.tsx:9]: This is coming in #1515, waiting on @sonechca to rebase and get that in.

File: src/frontend/next/src/components/Posts/LoadAutoScroll.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/LoadAutoScroll.tsx:45]: Need to fix the eslint warning you have on `$buttonRef.current`.  cc @birtony, who just fixed this kind of thing for #1227.

File: src/frontend/next/src/components/Posts/Posts.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Posts.tsx:1]: This file can be renamed to `src/frontend/next/src/components/Posts/index.tsx`

File: src/frontend/next/src/components/Posts/Posts.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Posts.tsx:12]: Why not use a proper Post type from our interfaces?

**Review by humphd — CHANGES_REQUESTED** [review]
Thanks for jumping on these fixes, we really need this PR to land soon, so the new frontend can be tested with our backend data.

**Comment by rogercyyu** [comment]
> 
> 
> Thanks for jumping on these fixes, we really need this PR to land soon, so the new frontend can be tested with our backend data.

Your welcome. Happy to help!

**Comment by humphd** [comment]
@rogercyuu is this ready for re-review then?

**Comment by rogercyyu** [comment]
@humphd yup

File: src/frontend/next/src/components/Post/PostComponent.tsx
Comment (humphd) [inline src/frontend/next/src/components/Post/PostComponent.tsx:1]: Rename this file `Post/index.tsx` vs `Post/PostComponent.tsx`

**Review by humphd — CHANGES_REQUESTED** [review]
One small fix, and this looks like it's ready.

I haven't tested the code, only read it.  Get another reviewer to take it for a spin.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Timeline.tsx:4]: Drop the `/index`, you can import `../../interfaces` and it will resolve to `interfaces/index`

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Timeline.tsx:7]: Later we can switch this to use `src/config` directly instead of the hook.  This is fine for now.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/Timeline.tsx:42]: `:any` isn't ideal.  Does it need typing (can it not infer it here)?

Not critical, but we could improve here or in follow-up.

File: src/frontend/next/src/components/Posts/index.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/index.tsx:11]: No `/index`

**Review by humphd — CHANGES_REQUESTED** [review]
Some small things, but this is really good.

Thank you for all this hard work, it's appreciated.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: It seems like typescript doesn't like it without `:any`:
image
Not sure what to do.

File: src/frontend/next/src/components/Post/index.tsx
Comment (tonyvugithub) [inline src/frontend/next/src/components/Post/index.tsx:1]: I feel like you can just move this Posts.tsx component inside the Posts folder.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (tonyvugithub) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: @rogercyyu : I would say keep any for now.

**Review by tonyvugithub — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (tonyvugithub) [inline src/frontend/next/src/components/Posts/Timeline.tsx:24]: @rogercyyu : This should be `theme.palette.primary.main`. I know the original file has it this way but TS complained about it I believe

**Review by tonyvugithub — CHANGES_REQUESTED** [review]

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Posts/Timeline.tsx:24]: good catch

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (manekenpix) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: I think the type here is
```ts
const postsTimeline = pages.map((page: Post[]): JSX.Element[] | JSX.Element =>
...
```

We should avoid using `any`. The whole point of using typescript is having type checking, but using `any` defeats that purpose.

**Review by PedroFonsecaDEV — DISMISSED** [review]
Well Done!

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: Problem is for `JSX.Element[]`, I get `'JSX' is not defined.eslintno-undef` unless we can use `// eslint-disable-next-line no-undef` to get rid of that.
image
hmmm...

File: src/frontend/next/src/components/Posts/index.tsx
Comment (PedroFonsecaDEV) [inline src/frontend/next/src/components/Posts/index.tsx:1]: This file should be named Posts.tsx

**Review by PedroFonsecaDEV — CHANGES_REQUESTED** [review]

**Comment by PedroFonsecaDEV** [comment]
@rogercyyu, how did you tested the PR?
We are experiencing some issues regarding the infinity scroll.

**Comment by rogercyyu** [comment]
@PedroFonsecaDEV hmmm works fine on my end. Not sure what's happening.

**Comment by PedroFonsecaDEV** [comment]
> @PedroFonsecaDEV hmmm works fine on my end. Not sure what's happening. I used the `npm run develop:next`. Is the build or serve one not working?

The infinity scroll starts an infinite loop after 4+ posts loads.

File: src/frontend/next/src/components/Posts/index.tsx
Comment (PedroFonsecaDEV) [reply src/frontend/next/src/components/Posts/index.tsx:1]: Okay. So keep it that way. I will fix something on the main page, and then I will relaunch CI/CD so we can see it on the netlify deployment.

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (tonyvugithub) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: @rogercyyu : Can you try sth like this Array<ReactElement>?

Comment (tonyvugithub) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: ReactElement basically JSX.Element

**Comment by humphd** [comment]
I want to land this today, what's missing that needs to happen?  If there's a scrolling bug, let's file that for follow-up vs holding this up.  Please update ASAP.

**Comment by humphd** [comment]
Looks like a rebase is needed for sure...

File: src/frontend/next/src/components/Posts/Timeline.tsx
Comment (rogercyyu) [reply src/frontend/next/src/components/Posts/Timeline.tsx:42]: @tonyvugithub thanks, that fixed it.

File: src/frontend/next/src/components/AdminButtons.tsx
Comment (humphd) [inline src/frontend/next/src/components/AdminButtons.tsx:3]: FYI: @rogercyyu and @sonechca we just need to co-ordinate this.  I think it's fine to land this now, but if the AdminButtons go in first (#1515), then let's rebase and remove this.

**Review by humphd — APPROVED** [review]
I think this is good for 1.5, and we can fix anything that comes of it in follow-ups.

**Review by manekenpix — APPROVED** [review]

File: src/frontend/next/src/components/Posts/index.tsx
Comment (humphd) [inline src/frontend/next/src/components/Posts/index.tsx:41]: When I load https://deploy-preview-1504--telescope-next.netlify.app/ it looks like `telescopeUrl` is undefined on Netlify, it is loading this URL:

https://deploy-preview-1504--telescope-next.netlify.app/undefined/posts?page=1

Filed #1589 to deal with this.

**Pull request merged**
